### ISSUE TITLE
Refactor entity manipulation

### ISSUE THREAD

**Issue by fpagnoux** [openfisca/openfisca-core#408]
## Problem

A significant part of the complexity in formulas comes from entities manipulation. 

To take a simple example, let's say I need to get the sum of all salaries in each family of my simulation, in order to determine if they can get some social prestation.

Today my code would be :

``` py
def function(self, simulation, period):
    plafond = 2000
    eligibilite = simulation.calculate('eligibilite')

    salaire_holder = simulation.compute('salaire', period)
    somme_salaires = self.sum_by_entity(salaire_holder)

    return period, eligibilite * (somme_salaires < plafond)
```

Several aspects are problematic here:
- We introduce the complex concept of `holder`. This is a leaky abstraction: unless you are an expert of `core`, you should not care about what is a holder.
- It is hard to stay serious when explaining to new contributors the subtle difference between `calculate` and `compute`.
- It is odd to use `self.something`: the computation doesn't actually have much to do with the formula itself, we are just summing a quantity for family members.

This example was the simplest case: some other functions meaning can be hard to grasp, such as `self.cast_from_entity_to_role`.
## Let's discuss a solution (I need you here)

I did some experiments and I think we can build a more **simple**,  and **complete** API to handle entities. Now I'd like to make sure we choose a syntax  as simple and natural as possible. 

So, for people who have been in touch with openfisca (or not) **how would you express the previous example in a way that seems the most natural to you** ?

Some propositions:

1:

``` py
    salaire_individus = simulation.calculate('salaire', period)
    somme_salaires = simulation.sum_in_entity(salaire_individus, Familles)

OR

    somme_salaires = simulation.sum_in_entity('salaire', period, Familles)
```

2:

``` py
    salaire_individus = simulation.calculate('salaire', period)
    somme_salaires = Familles.sum(simulation, salaire_individus)

OR

    somme_salaires = Familles.sum(simulation, 'salaire', period)
```
- Advantage: No need for the heavy `in_entity`
- Inconvenient: I don't find natural to give simulation as a first argument. It is however needed to know the composition of the families.

3

``` py
    salaire_individus = simulation.calculate('salaire', period)
    somme_salaires = Familles.sum(salaire_individus)

# The one-liner solution seems impossible here, as we need a reference to the simulation to run the calculations.
```
- Advantage : Compact and clear
- Inconvenient: Would only work if the result of simulation.calculate('') is not just a numpy array but something that countains a reference to the simulation. Not realistic today, but could be after @michelbl is done with his work on core.

4

```
YOUR SUGGESTION HERE
```

@edarin @SophieIPP @pierrehvz @adrienpacifico @benjello @laem @MattiSG @jdesboeufs @cbenz:  your opinions would be really valuable. **Especially** if you have not spend a lot of times on openfisca, as it gives you a less biased point of view ;).

**Comment by benjello** [comment]
@MalkIPP @MarionIPP @thomasdouenne your input is also valuable !

**Comment by benjello** [comment]
I like 3 very much. But keep in mind that you need to be able to do partial sum on some entity roles (I imagine taht could be easily handled by adding some arguments to the method sum).

**Comment by fpagnoux** [comment]
Yes, there would be in any case an extra role parameter to apply the sum -or any other operation- only on specific roles.

**Comment by cbenz** [comment]
Indeed @michelbl did a refactoring (WIP) [here](https://github.com/openfisca/openfisca-core/pull/402) which removes (and not just hides) the concept of `holder`.

To establish a parallel between our works, when we abtracted the computation as a syntax tree (AST) we introduced an operation `ValueForEntity(value, entity) -> new_value`.

My proposition:

```python
    salaire_individus = simulation.calculate('salaire', period)
    somme_salaires = salaire_individus.sum_by('familles')
    somme_salaires_chef = salaire_individus.sum_by('familles', role='chef')
```

(I'm not fond of `Familles` global object)

**Comment by adrienpacifico** [comment]
Naive question is there a sub-choices of two-lines vs. one-liner ?
Or do you plan to implement it such that both are possible ?

**Comment by fpagnoux** [comment]
The two lines solution is necessary IMO, if the quantity you want sum is not the direct calculation of another variable. For instance
```
salaire = simulation.calculate('salaire')
age = simulation.calculate('age')
eligible = (salaire < 1000) * (age > 18)
famille_eligible = Familles.sum(eligible)
```

The one liner is optional syntactic sugar, do you think it would be useful ?

**Comment by adrienpacifico** [comment]
I do agree for the two line formula, my concern was also on conditions.
I like the sugar syntax, I think reading/understanding formula would be easier.

**Comment by MattiSG** [comment]
IMO the point is how much do we want to make entities a first-class concept. If we do want to keep them and have contributors interact with them, I recommend presenting them as globals that would expose static methods to work with, leading to options 2 or 3.
Obviously, if the `salaire_individus` is anyway tied to the `simulation` and wouldn't work without it, there's no point in passing the `simulation` as a separate parameter, and the computation result should package its own `simulation` reference, leading to option 3.

As for the one liners: don't add them now, and don't design for them if they are sugar. I strongly recommend to design _basic_, _straightforward_ and **orthogonal** APIs. That is, each method should minimise both its arity and its functional overlap with other methods. Then, you might combine them in helpers over time, but these should **not** be part of the core. It's the UNIX way applied to APIs, and it's the simplest way that I know of to make sure there's a canonical way to make an operation, which means unified codebase, better learning curve, and simpler change.

**Comment by fpagnoux** [comment]
I'm not sure I understand what you mean by having helpers that are not in the core. What do you exactly mean by core ? `openfisca_core` package or something more specific ? 
Let's say we conclude that this one-liner would help us simplify formulas a lot. Where would add it ? In a different *file* than the basic API ? In a different *package* ?

**Comment by MattiSG** [comment]
Sorry, I was making a more general statement. “core” as in “don't design for it”. You need to focus on a set of basic, solid APIs first, even if they end up a bit cumbersome as long as all intents are clear. Then, if you add sugar on them and decide to package them as standard stuff, you might very well expose them in `openfisca_core`. But the software will be much more reliable and evolutive if the abstractions are built in layers.
If you present here a choice between options and present the “how it could be after yet another evolution step”, you're taking the risk that people vote on the very last line because it is sexy, while the previous lines combination is suboptimal.

**Comment by adrienpacifico** [comment]
What are the pros and cons of a `Familles` global object, what would it contains ?

I would go for 1) with keys. My arguments are :

- that it should prevent from importing `Entities` (or avoiding implicit global declarations when loading a simulation)
-  that it use only pure numpy array (no exotic complexities)
- Unexplainable personal taste...

But I'm not sure to really weight what is a stake between 1, 2, and 3...

To be explicit, I like:
```
salaire_individus = simulation.calculate('salaire', period)
somme_salaires = simulation.sum_in_entity(salaire_individus, 'famille', role = [?])
```

I think Christophe solution `salaire_individus.sum_by_familles()`,  if I think about my personal potential usage, may add complexity and withdraw flexibility. Let's say I want to compare income inequality over menage and famille. I could do:

```
for entity in ["famille", "menage"]:
    somme_sal_net = simulation.sum_in_entity('salaire_net', entity) 
    print some_inequality_func(somme_sal_net)
```

Many thanx to you all, and I hope I do not pollute the debate with mistakes and stupid interventions ;-) .

**Comment by MattiSG** [comment]
Counter-argument on `'famille'` (option :one:): passing an enum as a string is always a bad idea. Making entities types make everything clearer.

**Comment by adrienpacifico** [comment]
What do you mean by enum ? 
I don't know how future core will work but for now entities are not enum, it's a dict of objects right  ?
`Famille = simulation.entity_by_key_singular["famille"]`

To my understanding roles are kind of enum passed as string, wrong ?

**Comment by MattiSG** [comment]
What I mean is: how should OpenFisca react if I write `simulation.sum_in_entity(salaire_individus, 'herp')`?
If there is a limited set of allowed strings as the second parameter, then it’s an enum, not a string. And it would be better for the programming model to reflect this constraint.

**Comment by fpagnoux** [comment]
The `Familles` object would just contain a few information about the general structure of a familles (mainly the different roles).
In solutions 2 and 3, it would also offer a few methods to handle entities conversions (sum, max, min, any, all)...
Using a string as a parameter doesn't seem good to me. It is less user-friendly (possible mistakes, needs for quotes) and confusing, as we are not expecting any string but one of the few entities defined.

**Comment by fpagnoux** [comment]
By the way, I don't think the one-liner `somme_salaires = Familles.sum('salaire', period)` is possible here, as we would not have any reference to the simulation to run the computation. I'm adjusting the first post.

**Comment by adrienpacifico** [comment]
Thank you both for you answers and patience. I am voting for 1 then !

**Comment by fpagnoux** [comment]
>```py
salaire_individus = simulation.calculate('salaire', period)
somme_salaires = salaire_individus.sum_by_familles()
```

@cbenz, I didn't think about giving the `sum` method to the computation result `salaires_individus`, but it's an interesting option. About `sum_by_familles`, it's a little tricky: in our model, `Familles` is something defined for a given country. Adding this country-specific method to an object from the core doesn't seem obvious to me.

As @MattiSG said, I think the question is to know whether `Familles` or `Individus` is something our formula developers should know about. This was a key point in our discussion yesterday, about experimentation on a new openfisca core. But as of today, I don't see any incremental change that would hide the entities from the developer. So they would know about `Familles`, etc., because they declare it for each formula. From this, it seems IMO more relevant to have the developers interact with the entities than looking for the specific function `sum_in_familles`.

**Comment by laem** [comment]
```python
salaire_individus = simulation.calculate('salaire', period)
```

This line is very confusing to me. 
It's like we're defining a selector on the entity Individu without ever mentioning it. I guess 'salaire' is definend on Individu in its code and if there were a collision (e.g. Contrat.salaire), we would call it salaire_individu.

It may still be OK for the small set of entities that we have now though...

This is how I'd write the formula without any language constraint (pseudo code): 


```yaml
Variable xxx
    domain: Famille

    formula:
        and:
            Famille.eligibilite
            sum(Famille.individus.salaire) < plafond
```
We might even go further and not specify `Famille.` since it is obvious in the context of this variable. 

The little bit of magic here is the part `individus.salaire` because here '.salaire' is sugar for 'individus.map(individu => individu.salaire)'. I'm not sure the python style should be replaced by a functional one. I'm in favor of a very natural syntax for these types of operations on entities.

Going even further, accessing a property of a list could implicitly return the sum of the properties of each item of the list when the comparison is done against a scalar, here `plafond` :

```yaml
Variable xxx
    domain: Famille

    formula:
        and:
            eligibilite
            individus.salaire < plafond
```

The domain Entity is this only way to access variables (hence we omit it here). They are unambiguous (always namespaced by their entity).


Sorry if that's too far from the current code 😀

**Comment by fpagnoux** [comment]
I actually thought about something a little similar. Allowing myself more distance with the current openfisca, I would write: 

```py
def function(famille, period):
    plafond = 2000
    salaires = famille.individus.calculate('salaire')
    eligibilite = famille.calculate('eligibilite')
    somme_salaires = famille.sum(salaires)

    return eligibilite * (somme_salaires < plafond)
```

What I like, with both this and what you suggests, is that, from the developer point of view, we forget about this abstract concept of a simulation. We are coding a formula for person, or for a family, or for a company. We compute the salary of **a person**, not the one of a simulation.

This seems harder to implement, though...

**Comment by fpagnoux** [comment]
What do you think about the last suggestion (4) ?
See how formulas would be written in the [test country](https://github.com/openfisca/openfisca-core/blob/hide-simulation/openfisca_core/tests/test_countries.py#L68)

Our example case could be treated as in my previous post, or in "retro" mode:

4-retro
```py
def function(self, simulation, period):
    plafond = 2000
    salaires_individus = simulation.calculate('salaire')
    eligibilite = simulation.calculate('eligibilite')
    somme_salaires = simulation.famille.sum(salaires)

    return eligibilite * (somme_salaires < plafond)
```

**Comment by MattiSG** [comment]
Fixed by #415.
